cpp/src/join/joining.cu

```diff
@@ -253,260 +220,345 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
-    int left_table_end = num_left_cols - num_cols_to_join;
-    int right_table_begin = num_left_cols;
-
-    //create left and right output column data buffers
-    for (int i = 0; i < left_table_end; ++i) {
-        gdf_column* col = lnonjoincol[i];
-        gdf_column_view_augmented(result_cols[i], nullptr, nullptr, join_size,
-                                  col->dtype, 0, col->dtype_info, col->col_name);
-        int col_width = cudf::byte_width(*(result_cols[i]));
-        RMM_TRY( RMM_ALLOC((void**)&(result_cols[i]->data), col_width * join_size, 0) ); // TODO: non-default stream?
-        RMM_TRY( RMM_ALLOC((void**)&(result_cols[i]->valid), sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size), 0) );
-        CUDA_TRY( cudaMemset(result_cols[i]->valid, 0, sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size)) );
-        CHECK_STREAM(0);
-    }
-    for (int i = right_table_begin; i < result_num_cols; ++i) {
-        gdf_column* col = rnonjoincol[i - right_table_begin];
-        gdf_column_view_augmented(result_cols[i], nullptr, nullptr, join_size,
-                                  col->dtype, 0, col->dtype_info, col->col_name);
-        int col_width = cudf::byte_width(*(result_cols[i]));
-        RMM_TRY( RMM_ALLOC((void**)&(result_cols[i]->data), col_width * join_size, 0) ); // TODO: non-default stream?
-        RMM_TRY( RMM_ALLOC((void**)&(result_cols[i]->valid), sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size), 0) );
-        CUDA_TRY( cudaMemset(result_cols[i]->valid, 0, sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size)) );
-        CHECK_STREAM(0);
-    }
-    //create joined output column data buffers
-    for (int join_index = 0; join_index < num_cols_to_join; ++join_index) {
-        int i = left_table_end + join_index;
-        gdf_column* col = left_cols[left_join_cols[join_index]];
-        gdf_column_view_augmented(result_cols[i], nullptr, nullptr, join_size,
-                                  col->dtype, 0, col->dtype_info, col->col_name);
-        int col_width = cudf::byte_width(*(result_cols[i]));
-        RMM_TRY( RMM_ALLOC((void**)&(result_cols[i]->data), col_width * join_size, 0) ); // TODO: non-default stream?
-        RMM_TRY( RMM_ALLOC((void**)&(result_cols[i]->valid), sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size), 0) );
-        CUDA_TRY( cudaMemset(result_cols[i]->valid, 0, sizeof(gdf_valid_type)*gdf_valid_allocation_size(join_size)) );
-        CHECK_STREAM(0);
+    // Update first set of type and type infos from left
+    std::vector <gdf_dtype> rdtypes = cudf::column_dtypes(left);
+    std::vector <gdf_dtype_extra_info> rdtype_infos = cudf::column_dtype_infos(left);
+    std::vector<gdf_column*> left_nonjoin_column(left_nonjoin_indices.size());
+    std::vector<gdf_column*> right_nonjoin_column;
+
+    // Gathering all the left table columns not in joining indices 
+    std::transform(std::cbegin(left_nonjoin_indices), std::cend(left_nonjoin_indices),
```

Comment:
A function in `table` to return a table of only the specified columns would also eliminate a lot of redundancy:
```
/**---------------------------------------------------------------------------*
 * @brief Returns a new `table` containing the set of specified columns.
 * 
 * @param column_indices Indices of the desired columns
 * @return table New table containing only the desired columns
*---------------------------------------------------------------------------**/
table table::select(std::vector<gdf_size_type> const& column_indices){
    CUDF_EXPECTS(column_indices.size() < num_columns(), "Requested too many columns.");
    std::vector<gdf_column*> desired_columns;
    for(auto index : column_indices){
        desired_columns.push_back(get_column(index));
    }
    return table{desired_columns};
}
```

Reply:
@jrhemstad is this already available or should I introduce? I couldn't find it in latest 0.10

Reply:
Should I also introduce an offset variable which can offset the index ?

Reply:
`table::select` doesn't exist, I just prototyped it for the review.

>Should I also introduce an offset variable which can offset the index ?

Is this necessary? Seems to complicate an otherwise simple function. I'd think the user should be able to easily construct the `column_indices` with the desired columns.

Reply:
In my case for the `result` table, the right table will have a shist in index due to concatinattion, so I have to update the column indices by that offset to use this function, else do it using transform. If we have an offset, we can use offset to provide this information. So I was just curious whether it will have any additional benefits.

Reply:
I think its more appropriate to update the indices via a `transform` before passing them into `select` rather than adding an `offset`.

Reply:
ok